File: core/src/main/java/org/aresclient/ares/api/instrument/Instrument.java
package org.aresclient.ares.api.instrument;

import org.aresclient.ares.api.Ares;
import org.aresclient.ares.api.setting.Setting;

import java.util.ArrayList;
import java.util.List;

public abstract class Instrument {
    private final String name;
    private final String description;

    protected final Setting.Map<?> settings;
    protected final List<Component<?>> components = new ArrayList<>();

    public Instrument(String name, String description, Setting.Map<?> parentSettings) {
        this.name = name;
        this.description = description;
        this.settings = parentSettings.addMap(name);
    }

    public void tick() {
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    void addComponent(Component<?> component) {
        components.add(component);
        if(component instanceof Component.Settings<?> settingsComponent)
            settingsComponent.setSettings(this.settings.addMap(settingsComponent.getPathName()));
    }

    public void registerEvents() {
        Ares.getEventManager().register(this);
        Ares.getEventManager().register(getClass());

        for(Component<?> component : components) {
            if(component instanceof Component.Listener<?,?>) {
                Ares.getEventManager().register(component);
                Ares.getEventManager().register(component.getClass());
            }
        }
    }

    public void unregisterEvents() {
        Ares.getEventManager().unregister(this);
        Ares.getEventManager().unregister(getClass());

        for(Component<?> component : components) {
            if(component instanceof Component.Listener<?,?>) {
                Ares.getEventManager().unregister(component);
                Ares.getEventManager().unregister(component.getClass());
            }
        }
    }
}

File: src/main/kotlin/org/aresclient/ares/api/instruments/Instrument.kt
package org.aresclient.ares.api.instruments

import net.minecraft.client.util.math.MatrixStack
import org.aresclient.ares.api.Wrapper
import org.aresclient.ares.api.nrender.drawer.HudDrawer
import org.aresclient.ares.api.nrender.drawer.WorldDrawer
import org.aresclient.ares.api.setting.MapSetting

abstract class Instrument(val settings: MapSetting): Wrapper {
	private val components = ArrayList<Component<*>>()

	internal fun addComponent(component: Component<*>) {
		components.add(component)
	}

	abstract fun isListening(): Boolean

	open fun registerEvents() {
		if(!isListening()) return // TODO: MIGHT NOT NEED THIS

		EVENTS.register(this)
		EVENTS.register(javaClass)

		components.forEach {
			EVENTS.register(it)
			EVENTS.register(it.javaClass)
		}
	}

	open fun unregisterEvents() {
		if(isListening()) return

		EVENTS.unregister(this)
		EVENTS.unregister(javaClass)

		components.forEach {
			EVENTS.unregister(it)
			EVENTS.unregister(it.javaClass)
		}
	}

	/* ---------------------------------------------------------------------- */

	fun tick() {
		if(isListening()) onTick()
		components.forEach(Component<*>::tick)
	}

	fun motion() {
		if(isListening()) onMotion()
		components.forEach(Component<*>::motion)
	}

	fun renderHud(drawer: HudDrawer, matrixStack: MatrixStack, delta: Float) {
		if(isListening()) onRenderHud(drawer, matrixStack, delta)
		components.forEach { it.renderHud(drawer, matrixStack, delta) }
	}

	fun renderWorld(drawer: WorldDrawer, delta: Float) {
		if(isListening()) onRenderWorld(drawer, delta)
		components.forEach { it.renderWorld(drawer, delta) }
	}

	/* ---------------------------------------------------------------------- */

	protected open fun onTick() {}
	protected open fun onMotion() {}
	protected open fun onRenderHud(drawer: HudDrawer, matrixStack: MatrixStack, delta: Float) {}
	protected open fun onRenderWorld(drawer: WorldDrawer, delta: Float) {}
}
